server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -467,10 +488,27 @@ public void uploadMetadata(
             );
             try {
                 IndexOutput indexOutput = storeDirectory.createOutput(metadataFilename, IOContext.DEFAULT);
+                List<SegmentCommitInfo> infos = segmentInfosSnapshot.asList();
+                Map<String, Version> segmentToLuceneVersion = new HashMap<>();
+                for (SegmentCommitInfo segmentCommitInfo : infos) {
+                    SegmentInfo info = segmentCommitInfo.info;
+                    Set<String> segFiles = info.files();
+                    for (String file : segFiles) {
+                        segmentToLuceneVersion.put(file, info.getVersion());
+                    }
+                }
+
                 Map<String, String> uploadedSegments = new HashMap<>();
                 for (String file : segmentFiles) {
                     if (segmentsUploadedToRemoteStore.containsKey(file)) {
-                        uploadedSegments.put(file, segmentsUploadedToRemoteStore.get(file).toString());
+                        UploadedSegmentMetadata metadata = segmentsUploadedToRemoteStore.get(file);
+                        if (segmentToLuceneVersion.containsKey(metadata.originalFilename)) {
+                            metadata.setWrittenBy(segmentToLuceneVersion.get(metadata.originalFilename));
```

Comment:
is it possible to set writtenBy where `UploadedSegmentMetadata` is initialized?

Reply:
That is what I wanted to do initially. But the point when we create the `UploadedSegmentMetadata` (see below), we don't have access to the `SegmentInfo`, unless we add a parameter (that carries the version) and introduce a new overloaded `copyFrom( )`. We would need to add the new parameter in the signature of multiple functions.

https://github.com/opensearch-project/OpenSearch/blob/4b4d84e0bc663f8374ee63aea6dab4b8b159bf5b/server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java#L408-L419

So I favoured using a setter instead. Is there a better approach we could go for? That would surely help streamline this more.

Thanks!

Reply:
Yeah, sending a new parameter in copyFrom chain would not be good. Can we refactor a bit and move the initialization of `UploadedSegmentMetadata` elsewhere(maybe in ``) to make it easier to initialize it and modify in future. metadata is something i suspect we might change more frequently.

But don't keep this as a blocker, if you find a good refactoring approach we can do it separately as well.

Reply:
Hey @linuxpi, 
Marking this thread as resolved for now because we don't have a better way to do this currently. I will submit a new PR if I find a better way to do this.
Thanks!